submissions/Nemooochka/memory-pair/scripts.js

```diff
@@ -0,0 +1,142 @@
+    });
+
+    cardsWrapBlock.appendChild(cards);
+}
+
+function cardHover(event) {
+    const targetElem = event.target;
+
+    // setTimeout for Call Stack, execute after all previous code is done
+    setTimeout(function() {
+        if (targetElem.classList.contains("front")) {
+            // clear timer for resetting all hovered cards when user move to new card
+            clearTimeout(timerReset);
+
+            // show hovering of card smoothly
+            timerMouseOver = setTimeout(function () {
+                showedCard++;
+
+                // when 2 cards are shown and you move to the third, hide previous two
+                if (showedCard > 2) {
+                    showedCard = 1;
+                    prevCard = prevCardId = undefined;
+                    resetActiveCards();
+                }
+
+                currentCard = targetElem.closest('.flipper');
+                currentCard.classList.add('hover');
+                currentCardId = currentCard.getAttribute('data-pair');
+                currentCardIndex = currentCard.getAttribute('data-card');
+
+                if (currentCardId === prevCardId) {
+                    foundPairs++;
+                    // hide pairs smoothly
+                    setTimeout(function () {
+                        currentCard.closest('.flip-container').style.opacity = '0';
+                        prevCard.closest('.flip-container').style.opacity = '0';
+                    }, 500);
+
+                    // when all pair are founded
+                    if (foundPairs === cards.length) {
+                        foundPairs = 0;
+                        setTimeout(function () {
+                            cardsWrapBlock.innerHTML = '' +
+                                '<div class="congratulations">' +
+                                '<p>Wow! You are really good at it. Congratulations! :)</p>' +
+                                '<button class="btn btn-new_game">One more time?</button>' +
+                                '</div>';
+                            newGameBtn = cardsWrapBlock.querySelector('.btn-new_game');
+                            newGameBtn.addEventListener("click", () => {
+                                cardsWrapBlock.innerHTML = '';
+                                renderCards(Shuffle(pairArray));
+                            });
+                        }, 500);
+                    }
+                } else {
+                    prevCard = currentCard;
+                    prevCardId = currentCardId;
+                    prevCardIndex = currentCardIndex;
+                }
+            }, 500);
+
+            // reset hovered cards after 3 seconds if user doesn't move to another card.
+            timerReset = setTimeout(function () {
+                showedCard = 0;
+                prevCard = prevCardId = prevCardIndex = currentCard = currentCardId = currentCardIndex = undefined;
+                resetActiveCards();
+            }, 3000);
```

Comment:
If I change to const, get an error - Cannot access 'timerReset' before initialization.
Because at the top of cardHover function there is a line clearTimeout(timerReset); - to reset previous timerReset.

That's why initialized let timerReset at the top of scripts file.

Reply:
@MaxKalashnyk, If I change to const, get an error - Cannot access 'timerReset' before initialization.
Because at the top of cardHover function there is a line clearTimeout(timerReset); - to reset previous timerReset.

That's why initialized let timerReset at the top of scripts file.

Reply:
It's weird. You can declare them at the beginning of your code where you declared all of your constants. it will work because it will be in the global scope.

Reply:
@MaxKalashnyk, I get an error as the const variable should be declared immediately where declared and after I can just read, but not modify.
You mean declare as const timer = setTimeout() and after use as timer?

Reply:
I mean declare as const your time value `3000`

Reply:
I see that you used time value `300` for your timeouts three times. It's better to use simple constant for this 'magic number' and reuse it several times

Reply:
@MaxKalashnyk , oh, got it, thank you a lot. Sorry for the misunderstanding :)